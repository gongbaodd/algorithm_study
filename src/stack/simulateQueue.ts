import { Stack } from "./lib";

export class Queue<T> {
  a = new Stack<T>();

  b = new Stack<T>();

  private allToB() {
    while (!this.a.isEmpty()) {
      this.b.push(this.a.pop() as T);
    }
  }

  private allToA() {
    while (!this.b.isEmpty()) {
      this.a.push(this.b.pop() as T);
    }
  }

  enqueue(data: T) {
    this.allToA();
    this.a.push(data);
  }

  dequeue() {
    this.allToB();
    this.b.pop();
    this.allToA();
  }

  size() {
    return this.a.size() + this.b.size();
  }

  get tail() {
    this.allToA();
    return this.a.top();
  }

  get head() {
    this.allToB();
    return this.b.top();
  }
}
