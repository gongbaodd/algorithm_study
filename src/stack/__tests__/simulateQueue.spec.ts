import { Queue } from "../simulateQueue";

describe("simulate queue", () => {
  describe("empty queue", () => {
    it("should return top ==== null", () => {
      expect(new Queue().head).toBeNull();
    });

    it("should return tail ==== null", () => {
      expect(new Queue().tail).toBeNull();
    });

    it("should return size() ==== 0", () => {
      expect(new Queue().size()).toBe(0);
    });
  });

  describe("queue with data [1,2,3,4]", () => {
    it("queue function", () => {
      const q = new Queue<number>();
      q.enqueue(1);
      expect(q.tail).toBe(1);
      expect(q.head).toBe(1);
      expect(q.size()).toBe(1);

      q.enqueue(2);
      expect(q).toMatchSnapshot();
      expect(q.head).toBe(1);
      expect(q.tail).toBe(2);
      expect(q.size()).toBe(2);

      q.dequeue();
      expect(q.tail).toBe(2);
      expect(q.head).toBe(2);
      expect(q.size()).toBe(1);

      q.enqueue(3);
      expect(q.tail).toBe(3);
      expect(q.head).toBe(2);
      expect(q.size()).toBe(2);

      q.enqueue(4);
      expect(q.tail).toBe(4);
      expect(q.head).toBe(2);
      expect(q.size()).toBe(3);

      q.dequeue();
      expect(q.tail).toBe(4);
      expect(q.head).toBe(3);
      expect(q.size()).toBe(2);
    });
  });
});
