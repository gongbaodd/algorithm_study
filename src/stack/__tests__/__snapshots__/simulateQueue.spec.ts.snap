// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simulate queue queue with data [1,2,3,4] queue function 1`] = `
Queue {
  "a": Stack {
    "pHead": LNode {
      "data": null,
      "next": LNode {
        "data": 2,
        "next": LNode {
          "data": 1,
          "next": null,
        },
      },
    },
  },
  "b": Stack {
    "pHead": LNode {
      "data": null,
      "next": null,
    },
  },
}
`;
